### ISSUE TITLE
Source maps still don't seem to be working

### ISSUE THREAD

**Issue by tyvdh** [robertcepa/toucan-js#54]
Followed the guide in the README and the thread but I'm still getting errors when attempting to link to my source map.

<img width="780" alt="Screen Shot 2021-01-12 at 8 34 49 PM" src="https://user-images.githubusercontent.com/4383610/104394992-bc7e4c80-5515-11eb-9f26-1b49bfcd786c.png">

Maybe I'm missing or misunderstanding something still?

```
addEventListener('fetch', (event) => {
  const sentry = new Toucan({
    dsn: SENTRY_DSN,
    event,
    environment: 'development',
    release: 'pre-release',
    allowedCookies: /(.*)/,
    allowedHeaders: /(.*)/,
    allowedSearchParams: /(.*)/,
    rewriteFrames: {
      root: '/'
    }
  })

  event.respondWith(handleRequest(event, sentry))
})
```

```
const SentryWebpackPlugin = require('@sentry/webpack-plugin')

module.exports = {
  entry: './src/index.js',
  target: 'webworker',
  devtool: 'cheap-module-source-map',
  node: {
    fs: 'empty'
  },
  plugins: [
    new SentryWebpackPlugin({
      authToken: 'abc123',
      release: 'pre-release',
      org: 'abc',
      project: '123',
      include: './dist',
      urlPrefix: '/',
    }),
  ],
}
```

**Comment by robertcepa** [comment]
Confix seems fine to me. Have you tried devtool: source-map?

**Comment by tyvdh** [comment]
I have yes, I can't use that directly though as the output is too large to upload. 
```
✨  Built successfully, built project size is 1 MiB. ⚠️  Your built project has grown past the 1MiB size limit and may fail to deploy. ⚠️ 
```

Currently I'm getting around this with two different webpack files, one to build and upload the src map and one to just upload the worker code. Works fine.

**Comment by tyvdh** [comment]
```
// wrangler.toml

name = "name"
type = "webpack"
account_id = "account-id-abc123"
workers_dev = true
webpack_config = "webpack.config.js"

[env.srcmap]
webpack_config = "webpack.config.srcmap.js"
```

```
// webpack.config.js

const webpack = require('webpack')

module.exports = {
  entry: './src/index.js',
  target: 'webworker',
  node: {
    fs: 'empty'
  },
  module: {
    rules: [
      {
        test: /\.js$/, exclude: /node_modules/, loader: 'babel-loader'
      }
    ]
  },
  plugins: [
    new webpack.ExtendedAPIPlugin(),
  ]
}
```

```
// webpack.config.srcmap.js

const SentryPlugin = require('webpack-sentry-plugin')
const webpack = require('webpack')

module.exports = {
  entry: './src/index.js',
  target: 'webworker',
  devtool: 'source-map',
  node: {
    fs: 'empty'
  },
  module: {
    rules: [
      {
        test: /\.js$/, exclude: /node_modules/, loader: 'babel-loader'
      }
    ]
  },
  plugins: [
    new webpack.ExtendedAPIPlugin(),
    new SentryPlugin({
      apiKey: 'sentry-api-key-abc123',
      organization: 'org-abc123',
      project: 'project-abc123',
      deleteAfterCompile: true,
      release: (hash) => hash
    })
  ],
}
```

```
// package.json

{
  "private": true,
  "main": "./src/index.js",
  "scripts": {
    "start": "wrangler dev",
    "deploy": "wrangler build -e srcmap ; wrangler publish"
  },
  "dependencies": {
    ...
  },
  "devDependencies": {
    ...
  }
}
```

```
// index.js

...

const sentry = new Toucan({
  pkg,
  event,
  dsn: SENTRY_DSN,
  environment: isDev ? 'development' : 'production',
  release: __webpack_hash__,
  attachStacktrace: true,
  allowedCookies: /(.*)/,
  allowedHeaders: /(.*)/,
  allowedSearchParams: /(.*)/,
  rewriteFrames: {
    root: '/'
  }
})

...
```

**Comment by olizilla** [comment]
i've been wrestling with source-map generation today too, and I found that I could not get source maps to work with CF Workers + SentryWebpackPlugin + Toucan unless I removed the `//# sourceMappingURL=` link from the end of the js bundle by setting `devtool: hidden-source-map` in the webpack config.

We're using webpack 5, and wrangler 1.19 and have configured wrangler to treat the project as `type = "javascript"`.

We configure the SentryWebpackPlugin like this:

```
      new SentryWebpackPlugin({
        release: gitRevisionPlugin.version(),
        include: './dist',
        urlPrefix: '/',
        org: '***',
        project: process.env.SENTRY_PROJECT,
        authToken: process.env.SENTRY_TOKEN
      })
```

and Toucan like this:

```
new Toucan({
    dsn: SENTRY_DSN,
    context: event,
    allowedHeaders: ['user-agent'],
    allowedSearchParams: /(.*)/,
    debug: false,
    rewriteFrames: {
      root: '/'
    },
    environment: ENV,
    release: VERSION,
    pkg
  })
```
which means we generate the bundle js and the source map in the dist dir, and both files get uploaded to sentry and then cloudflare on publish. The source maps show up on the release in Sentry but they just would not work until I switched the webpack devtool from `devtool: source-map` `devtool: hidden-source-map`, which only removes `//# sourceMappingURL=` link from the bundle.

Is that expected? I'd love to understand what is going on there. From the README for toucan it seem like that is not normal.

**Comment by robertcepa** [comment]
@olizilla Do you have `.map` file along with your compiled `.js` file in your `dist` folder when you use `devtool: source-map`? If yes, and assuming your worker script is named `worker.js`, is your `worker.js` referencing the correct map file? It should look like this: ` sourceMappingURL=worker.js.map`. Also check if both `worker.js` and `worker.js.map` show in Sentry under your project's source maps (or release artifacts if you use older version of Sentry):
<img width="1438" alt="Screen Shot 2021-10-02 at 10 37 38 AM" src="https://user-images.githubusercontent.com/9921735/135726856-56217784-1eb7-4ec3-b006-976ea3d66488.png">

**Comment by xmflsct** [comment]
Seems also not working with modules.

**Comment by olizilla** [comment]
Aye, we just switched to esm modules on cloudflare, and our source maps are no longer showing up on Sentry.

**Comment by olizilla** [comment]
I got it working with esm by using the `rewriteFrames.iteratee` property to strip a leading `.` character from the `Stackframe.filename` that we were seeing on errors sent from cloudflare to sentry.

```js
rewriteFrames: {
      // strip . from start of the filename ./worker.mjs as set by cloudflare, to make absolute path `/worker.mjs`
      iteratee: (frame) => ({ ...frame, filename: frame.filename.substring(1) })
    },
```

**happy sourcemap'd sentry error from an es module flavour worker**

<img width="848" alt="Screenshot 2022-02-25 at 22 01 21 copy" src="https://user-images.githubusercontent.com/58871/155976912-c8f3d872-39b1-4d45-b549-8de0bb79ba2a.png">


**Before this change** the raw JSON view of an error on Sentry showed that the `filename` property would be set to `./worker.js`

**unhappy sentry errror**

```json
  "exception": {
    "values": [
      {
        "type": "Error",
        "value": "A deliberate error!",
        "stacktrace": {
          "frames": [
            {
              "function": "async Object.fetch",
              "filename": "./worker.mjs",
              "abs_path": "./worker.mjs",
```

If I configure Toucan with `rewriteFrames.root: '/'` to prepend `/` to the `filename`, then Sentry shows, the `filename` and  `abs_path` as `"/./worker.mjs"` and the only way to get Sentry to locate the uploaded bundle was to also set the `urlPrefix` option one uploading the files to Sentry to the same like `urlPrefix: ~/./` which seemed kinda gross, but worked... I opted for just stripping the `.` from `./worker.mjs` in the Toucan config as it seemed neater, as it also meant I could avoid specifying the `urlPrefix` when uploading them as the default is `~/` which works as, i learnt recently:

> as `~/` is wildcard matcher for any `protocol+host` pair.
– https://github.com/getsentry/sentry-cli/issues/894#issuecomment-944347277

**Comment by ptim** [comment]
> I got it working with esm by using the rewriteFrames.iteratee property to strip a leading . character from the Stackframe.filename that we were seeing on errors sent from cloudflare to sentry.

Thanks so much - this fixed the issue for me, too 🙏

I'm building with esbuild (so no webpack plugin as mentioned in docs), so it was additionally necessary for me to set  
`--urlPrefix '/'`

**Comment by olizilla** [comment]
Note that this broken _again_ when we upgraded to **wrangler v2**, and the fix was to

- Rename our worker bundle build output to `worker.js` (was previously `worker.mjs`)
- Stop stripping the initial char from the `stack.filename` in the Toucan config and revert back to setting `rewriteFrames.root: '/'` to prefix a `/` on the filename, leaving us with `/worker.js` as the reported `abs_path` in sentry.

see: https://github.com/web3-storage/web3.storage/pull/1593

**Comment by olizilla** [comment]
OK! It seems that Cloudflare does it's own bundling of your worker code. 

For your source map to work you need to disable that new feature in your wrangler.toml by setting `no_bundle = true`. 

🎩 Hat tip to @hugomrdias for the discovery in https://github.com/nftstorage/nft.storage/pull/2094

**Comment by olizilla** [comment]
We also had to revert the `rewriteFrames` config listed in https://github.com/robertcepa/toucan-js/issues/54#issuecomment-1054169127 back to the simpler `root: /` form, and update our bundle name to be `worker.js`

```js
rewriteFrames: {
      root: '/'
    },
```

see: https://github.com/web3-storage/web3.storage/pull/1838

**Comment by gobengo** [comment]
@robertcepa Do you expect `toucan-js@3.x` to be able to work when wrangler's `no_bundle` option is `false`?
*  I believe this setting is currently [required for projects using D1](https://developers.cloudflare.com/workers/wrangler/bundling/#d1-bindings).
* all the 3.x examples use `no_bundle=true` https://github.com/search?q=repo%3Arobertcepa%2Ftoucan-js%20no_bundle&type=code
* I upgraded to toucan-js@3.x here: https://github.com/web3-storage/w3up/pull/728/files#diff-26442b60298c1d3462516817b6339bc31abe23e6b7cb7acc34367ea02cfb9975R48
  * that alone didn't fix how source maps appear.
  * @olizilla just helped me realize that it could be because
    * that project still uses `no_bundle=false`
      * so the build step [sends `worker.js.map` source maps to sentry directly](https://github.com/web3-storage/w3up/blob/main/packages/access-api/scripts/cli.js#L92)
      * but on deploy to workers, a different source map might be generated, and `toucan-js` might be using that source 
map info

Would you expect `toucan-js@3.x` to support stack traces on errors sent from within cloudflare worker with `no_bundle=false`? I'm not sure if lack of examples is due to lack of support at all, or there just hasn't been a need yet to make an example.

**Comment by gobengo** [comment]
> This should be a solved problem in [3.0.0](https://github.com/robertcepa/toucan-js/releases/tag/toucan-js%403.0.0). Mind giving it a shot?

@robertcepa I updated our project to use `toucan-js@3.x` and emulated the [wrangler-basic](https://github.com/robertcepa/toucan-js/tree/master/examples/wrangler-basic) example (omitting wrangler.toml `no_bundle` i.e. `no_bundle=false`).
It seems to be working!

Thanks for all your advice. 🙏
